README.md
# Risk-TTS
The classic board game Risk for TableTop Simulator.


changelog.txt
Change Log
1.1 - Saves Update
  Added - 
    Games can now be saved!
      As there are alot of states the game can be in when saving, so I may have missed a few button updates. These issues will simply
        graphical issues. The game will still continue correctly as all crucial values ARE being stored. But please do let me know about
        these issues on the issue tracker.
    Function to check if an attack is being made over water.
      Needed for mission and special gamerule later.
    Tooltips to territory buttons with the territory name.
  Fixes -
    Attack Again button giving a minor error about not having a function it can call.
    Player's clock continued to tick after game over.
    Error when a player is seated in a color that is not assigned to the table.
    Incorrect players able to click attack again buttons on the dice roller boxes.

RiskDiceRoller.lua
function onLoad()
  trackedDice = {}
  startedRolling = false
  defender = false
  if self.guid == "44e410" then
    defender = true
  end
  currentUser = "Grey"
  tColor = {1,1,1}
  buttons = {}
  buttons[1] = {click_function = "none", index = 0, function_owner = self, label = "Attack Again?", position = {0, -2.14, -3.2}, rotation = {0, 180, 0}, scale = {0, 0, 0}, width = 2000, height = 400, font_size = 330, attackAgain = false}
  buttons[2] = {click_function = "repeatYes", index = 1, function_owner = self, label = "Yes", position = {0.54, -2.14, -3.6}, rotation = {0, 180, 0}, scale = {0, 0, 0}, width = 998, height = 400, font_size = 400, on = false}
  buttons[3] = {click_function = "repeatNo", index = 2, function_owner = self, label = "No", position = {-0.54, -2.14, -3.6}, rotation = {0, 180, 0}, scale = {0, 0, 0}, width = 998, height = 400, font_size = 400, on = false}
  
  if defender then
    bag = getObjectFromGUID("27e50f")
    dieOffset = 45
    buttons[4] = {click_function = "rollDice1", index = 3, function_owner = self, label = "1", position = {0.54, -2.14, -2.8}, rotation = {0, 180, 0}, scale = {0.5, 0.5, 0.5}, width = 998, height = 400, font_size = 400, on = false}
    buttons[5] = {click_function = "rollDice2", index = 4, function_owner = self, label = "2", position = {-0.54, -2.14, -2.8}, rotation = {0, 180, 0}, scale = {0.5, 0.5, 0.5}, width = 998, height = 400, font_size = 400, on = false}
  else
    bag = getObjectFromGUID("82056b")
    dieOffset = -45
    buttons[4] = {click_function = "rollDice1", index = 3, function_owner = self, label = "1", position = {0.7, -2.14, -2.8}, rotation = {0, 180, 0}, scale = {0.5, 0.5, 0.5}, width = 666, height = 400, font_size = 400, on = false}
    buttons[5] = {click_function = "rollDice2", index = 4, function_owner = self, label = "2", position = {0, -2.14, -2.8}, rotation = {0, 180, 0}, scale = {0.5, 0.5, 0.5}, width = 666, height = 400, font_size = 400, on = false}
    buttons[6] = {click_function = "rollDice3", index = 5, function_owner = self, label = "3", position = {-0.7, -2.14, -2.8}, rotation = {0, 180, 0}, scale = {0.5, 0.5, 0.5}, width = 666, height = 400, font_size = 400, on = false}
  end
  for k,v in pairs(buttons) do
    self.createButton(v)
  end
  map = getObjectFromGUID("be2f30")
  bag.interactable = false
  pColors = {
    White = {1,1,1},
    Red = {0.856, 0.1, 0.094},
    Orange = {0.956, 0.392, 0.113},
    Yellow = {0.905, 0.898, 0.172},
    Green = {0.192, 0.701, 0.168},
    Blue = {0.118, 0.53, 1},{0.25},
    Pink = {0.96, 0.439, 0.807},
    Grey = {0.5,0.5,0.5},
  }
  tColors = {
    White = {0.25, 0.25, 0.25, 1},
    Red = {1, 1, 1, 1}, 
    Orange = {1, 1, 1, 1},
    Yellow = {0, 0, 0, 1},
    Green = {1, 1, 1, 1},
    Blue = {1, 1, 1, 1}, 
    Pink = {1, 1, 1, 1},
    Grey = {1, 1, 1, 1},
  }
  diceAmount({0})
  Global.call('objectLoadFinished',{self.guid})
end
function getSaveData()
  local data = {
    currentUser = currentUser,
    trackedDice = trackedDice,
    startedRolling = startedRolling,
    attackAgain = buttons[1].attackAgain,
    buttonStates = {}
  }
  
  for i=1, #buttons - 3 do
    data.buttonStates[i] = buttons[i+3].on
  end
  return data
end
function loadSaveData(d)
  data = d[1]
  --addNotebookTab({title = "roller", body = JSON.encode_pretty(data)})
  currentUser = data.currentUser
  trackedDice = data.trackedDice
  startedRolling = data.startedRolling
  --update attack again button
  showAttackAgain({data.attackAgain,currentUser})
  --update the amount to roll buttons
  for i=1,#data.buttonStates do
    if data.buttonStates[i]  then
      buttons[i+3].color = pColors[currentUser]
      buttons[i+3].font_color = tColors[currentUser]
      buttons[i+3].on = true
    else
      buttons[i+3].color = pColors["Grey"]
      buttons[i+3].font_color = pColors["Grey"]
      buttons[i+3].on = false
    end
    self.editButton(buttons[i+3])
  end
end
function onUpdate()
  if #trackedDice > 0 and startedRolling then
    for k,v in pairs(trackedDice) do
      local obj = getObjectFromGUID(v)
      if obj.resting ~= true then
        return
      end
    end
    startedRolling = false
    doneRoll()
  end
end
function repeatNo(o,c) map.call('attackAgain',{false}) end
function repeatYes(o,c) map.call('attackAgain',{true}) end
function rollDice1(o,c) rollDice(1,c) end
function rollDice2(o,c) rollDice(2,c) end
function rollDice3(o,c) rollDice(3,c) end
function rollDice(num,c)
  if startedRolling then
    return false
  end
  if buttons[num+3].on == false then
    return
  end
  if c ~= currentUser then
    broadcastToColor("It is not your turn to roll the dice.",c,tColor)
    return
  end
  startedRolling = true
  diceAmount({0,currentUser})
  for i=1,num do
    local loc = {dieOffset + (math.random(200)/100 - 1), 35 +(i*2), 47 + (math.random(200)/100 - 1)}
    local params = {}
    params.position = loc
    params.callback = "callback"
    params.callback_owner = self
    bag.takeObject(params)
  end
end
function doneRoll(b,c)
  local str = ""
  if defender then
    str = "Defender Rolled - "
  else
    str = "Attacker Rolled - "
  end
  local values = {}
  for k,v in pairs(trackedDice) do
    local obj = getObjectFromGUID(v)
    table.insert(values,obj.getValue())
    obj.destruct()
  end
  table.sort(values,function(a,b) return a > b end)
  for k,v in ipairs(values) do
    str = str .. v .. " "
  end
  broadcastToAll(str,tColor)
  trackedDice = {}
  map.call('getDiceRolls',{defender, values})
end
function callback(o,p)
  table.insert(trackedDice,o.guid)
  --o.interactable = false
  o.setRotation({math.random(359),math.random(359),math.random(359)})
  o.roll()
end
function diceAmount(data)
  num = data[1]
  currentUser = data[2]
  for i=1,num do
    buttons[i+3].color = pColors[currentUser]
    buttons[i+3].font_color = tColors[currentUser]
    buttons[i+3].on = true
    self.editButton(buttons[i +3])
  end
  local numButtons = #buttons - 3
  if num < numButtons then
    local offset = numButtons - num
    for i=1,offset do
      buttons[num + i + 3].color = pColors["Grey"]
      buttons[num + i + 3].font_color = pColors["Grey"]
      buttons[num + i + 3].on = false
      self.editButton(buttons[num + i + 3])
    end
  end
end

function showAttackAgain(data)
  local activate = data[1]
  for i=1,3 do
    if activate then
      buttons[i].scale = {0.5,0.5,0.5}
      buttons[i].color = pColors[data[2]]
      if i == 1 then
        buttons[i].attackAgain = true
      end
    else
      buttons[i].scale = {0,0,0}
      if i == 1 then
        buttons[i].attackAgain = false
      end
    end
    self.editButton(buttons[i])
  end
end
function none()

end

RiskConfigController.lua
function onLoad()
  gameStarted = false
  buttons = {}
  buttons[1] = {click_function = "strategicConfig", function_owner = self, label = "Long Range Strategic Moves", tooltip = "Allows You to move units during the strategic move phase to any territory \nso long as there is a path through territories you control.",
                    position = {0.28, 0.53, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3240, height = 940, font_size = 200, alignment = 5, index = 1,value = false}
  buttons[2] = {click_function = "setTradeConfig", function_owner = self, label = "Turning in Sets after defeating \nan opponent and recieved cards.", 
                    position = {0.16, 0.53, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3240, height = 940, font_size = 200, alignment = 3, index = 3,value = false}
  buttons[3] = {click_function = "setTerritoryConfig", function_owner = self, label = "Gain Units for every territory show\nwhen trading in a set", 
                    position = {0.04, 0.53, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3240, height = 940, font_size = 200, alignment = 3, index = 5,value = false}
  buttons[4] = {click_function = "reshuffleConfig", function_owner = self, label = "Enable Reshuffle Cards When Empty", 
                    position = {-0.08, 0.53, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3240, height = 940, font_size = 200, alignment = 3, index = 7,value = false}

  buttonsOutline = {}
  buttonsOutline[1] = {click_function = "none", function_owner = self, label = "",
                    position = {0.28, 0.51, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3300, height = 1000, font_size = 200, alignment = 5, index = 0,value = false}
  buttonsOutline[2] = {click_function = "none", function_owner = self, label = "", 
                    position = {0.16, 0.51, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3300, height = 1000, font_size = 200, alignment = 3, index = 2,value = false}
  buttonsOutline[3] = {click_function = "none", function_owner = self, label = "", 
                    position = {0.04, 0.51, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3300, height = 1000, font_size = 200, alignment = 3, index = 4,value = false}
  buttonsOutline[4] = {click_function = "none", function_owner = self, label = "", 
                    position = {-0.08, 0.51, -0.82}, rotation = {0, 90, 0}, scale = {0.05, 0.05, 0.05}, width = 3300, height = 1000, font_size = 200, alignment = 3, index = 6,value = false}
 -- local data = {click_function = "INSERT_FUNCTION", function_owner = self, label = "Restart/Regen", 
 --                   position = {-0.0799999982118607, 0.509999990463257, 0.824}, rotation = {0, 90, 0}, scale = {0.0500000007450581, 0.0500000007450581, 0.0500000007450581}, width = 3300, height = 1000, font_size = 200, tooltip = "Restart/Regen", alignment = 3}
  for k,v in pairs(buttons) do
    if v.value == false then
      buttonsOutline[k].color = {0.856, 0.1, 0.094}
    else
      buttonsOutline[k].color = {0.192, 0.701, 0.168}
    end
    v.color = {1,1,1}
    self.createButton(buttonsOutline[k])
    self.createButton(v)
  end
  Global.call('objectLoadFinished',{self.guid})
end
function strategicConfig(o,c)
  if gameStarted then
    return
  end
  if buttons[1].value then
    buttons[1].value = false
    buttonsOutline[1].color = {0.856, 0.1, 0.094}
  else
    buttons[1].value = true
    buttonsOutline[1].color = {0.192, 0.701, 0.168}
  end
  self.editButton(buttonsOutline[1]) 
  self.editButton(buttons[1])
end
 
function setTradeConfig(o,c)
  if gameStarted then
    return
  end
  if buttons[2].value then
    buttons[2].value = false
    buttonsOutline[2].color = {0.856, 0.1, 0.094}
  else
    buttons[2].value = true
    buttonsOutline[2].color = {0.192, 0.701, 0.168}
  end
  self.editButton(buttonsOutline[2])
  self.editButton(buttons[2])
end

function setTerritoryConfig(o,c)
  if gameStarted then
    return
  end
  if buttons[3].value then
    buttons[3].value = false
    buttonsOutline[3].color = {0.856, 0.1, 0.094}
  else
    buttons[3].value = true
    buttonsOutline[3].color = {0.192, 0.701, 0.168}
  end
  self.editButton(buttonsOutline[3])
  self.editButton(buttons[3])
end

function reshuffleConfig(o,c)
  if gameStarted then
    return
  end
  if buttons[4].value then
    buttons[4].value = false
    buttonsOutline[4].color = {0.856, 0.1, 0.094}
  else
    buttons[4].value = true
    buttonsOutline[4].color = {0.192, 0.701, 0.168}
  end
  self.editButton(buttonsOutline[4])
  self.editButton(buttons[4])
end
function getConfigValues()
  local info = {}
  info.longRangeStrategic = buttons[1].value
  info.allowSetAfterEnemyDefeat = buttons[2].value
  info.giveAllUnitsOnSetTerritories = buttons[3].value
  info.reshuffleCardsOnDeckEmpty = buttons[4].value
  return info
end

function loadSaveData(config)
  if config.longRangeStrategic then strategicConfig() end
  if config.allowSetAfterEnemyDefeat then setTradeConfig() end
  if config.giveAllUnitsOnSetTerritories then setTerritoryConfig() end
  if config.reshuffleCardsOnDeckEmpty then reshuffleConfig() end
end

function toggleConfigButtons()
  if gameStarted then
    gameStarted = false
  else
    gameStarted = true
  end
end
function none()
  return
end

RiskController.lua
function onLoad()
  pColors = {
    White = {1,1,1},
    Red = {0.856, 0.1, 0.094},
    Orange = {0.956, 0.392, 0.113},
    Yellow = {0.905, 0.898, 0.172},
    Green = {0.192, 0.701, 0.168},
    Blue = {0.118, 0.53, 1},{0.25},
    Pink = {0.96, 0.439, 0.807},
    Grey = {0.5,0.5,0.5}
  }
  tColors = {
    White = {0.25, 0.25, 0.25, 1},
    Red = {1, 1, 1, 1}, 
    Orange = {1, 1, 1, 1},
    Yellow = {0, 0, 0, 1},
    Green = {1, 1, 1, 1},
    Blue = {1, 1, 1, 1}, 
    Pink = {1, 1, 1, 1},
    Grey = {1, 1, 1, 1},
  }
  data={}
  --data[1] = {index = 0, click_function = "doNothing",   function_owner = self, label = "Press Start", position = {-6, 0.6, -7.3}, scale = {1, 0.5, 1.6}, width = 2000, height = 400, font_size = 400, tooltip = "Turn"}
  data[1] = {index = 0, click_function = "doNothing", function_owner = self, label = "0", position = {2.1, 0.6, 7}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300, tooltip = "Amount Left Behind"}
  
  data[2] = {index = 1, click_function = "doNothing",   function_owner = self, label = "Pre\nGame", position = {-5.6, 0.6, -1.1}, scale = {1, 0.5, 2}, width = 2300, height = 900, font_size = 400, tooltip = "Phase"}
  data[3] = {index = 2, click_function = "endPhase",                 function_owner = self, label = "End phase", position = {5.9, 0.6, 0.1}, scale = {1, 0.5, 2}, width = 2000, height = 500, font_size = 400}
  data[4] = {index = 3, click_function = "doNothing",   function_owner = self, label = "Territory 1", position = {0.3, 0.6, -2.3}, scale = {1, 0.5, 2}, width = 3300, height = 500, font_size = 300}
  data[5] = {index = 4, click_function = "doNothing",   function_owner = self, label = "Territory 2", position = {0.3, 0.6, 0.1}, scale = {1, 0.5, 2}, width = 3300, height = 500, font_size = 300}
  data[6] = {index = 5, click_function = "confirm", function_owner = self, label = "Confirm", position = {5.9, 0.6, -2.3}, scale = {1, 0.5, 2}, width = 2000, height = 500, font_size = 400}
  data[7] = {index = 6, click_function = "doNothing",   function_owner = self, label = "0", position = {2.1, 0.6, 4.7}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300}
  data[8] = {index = 7, click_function = "addTen",      function_owner = self, label = "+ 10", position = {5.5, 0.6, 2.4}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300}
  data[9] = {index = 8, click_function = "addAll",        function_owner = self, label = "+ All", position = {7.2, 0.6, 2.4}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300}
  data[10] = {index = 9, click_function = "subTen",    function_owner = self, label = "- 10", position = {5.5, 0.6, 4.7}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300}
  data[11] = {index = 10, click_function = "subAll",      function_owner = self, label = "- All", position = {7.2, 0.6, 4.7}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300}
  data[12] = {index = 11, click_function = "doNothing", function_owner = self, label = "0", position = {2.1, 0.6, 2.4}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300}
  data[13] = {index = 12, click_function = "doNothing", function_owner = self, label = "Amount to Add", position = {-1, 0.6, 2.4}, scale = {1, 0.5, 2}, width = 2100, height = 500, font_size = 300}
  data[14] = {index = 13, click_function = "doNothing", function_owner = self, label = "Left to Place", position = {-1.0, 0.6, 4.7}, scale = {1, 0.5, 2}, width = 2100, height = 500, font_size = 300}
  data[15] = {index = 14, click_function = "doNothing", function_owner = self, label = "Push Start\nTo Begin", position = {-0.035, 0.6, -5.7}, scale = {1, 0.5, 2}, width = 7600, height = 1000, font_size = 500}
  data[16] = {index = 15, click_function = "card1", function_owner = self, label = "Card 1", position = {-5.6, 0.6, 2.4}, scale = {0, 0, 0}, width = 2300, height = 500, font_size = 250}
  data[17] = {index = 16, click_function = "card2", function_owner = self, label = "Card 2", position = {-5.6, 0.6, 4.7}, scale = {0, 0, 0}, width = 2300, height = 500, font_size = 250}
  data[18] = {index = 17, click_function = "card3", function_owner = self, label = "Card 3", position = {-5.6, 0.6, 7}, scale = {0, 0, 0}, width = 2300, height = 500, font_size = 250}
  data[19] = {index = 18, click_function = "doNothing", function_owner = self, label = "Left Behind", position = {-1, 0.6, 7}, scale = {1, 0.5, 2}, width = 2100, height = 500, font_size = 300}
  data[20] = {index = 19, click_function = "subOne", function_owner = self, label = "-", position = {3.8, 0.6, 4.7}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300, tooltip = "Subtract 1"}
  data[21] = {index = 20, click_function = "addOne", function_owner = self, label = "+", position = {3.8, 0.6, 2.4}, scale = {1, 0.5, 2}, width = 800, height = 500, font_size = 300, tooltip = "Add 1"}
  data[22] = {index = 21, click_function = "doNothing", function_owner = self, label = "Turn In Sets\nHere", position = {-5.6, 0.6, 4.7}, scale = {0, 0.5, 0}, width = 2300, height = 1600, font_size = 400}
  
  
  
  
  
  
  
  board = getObjectFromGUID("be2f30")
  for k,v in pairs(data) do
    self.createButton(v)
  end
  cardZone = getObjectFromGUID("1398dd")
  local _scale = {62.38, 1.0, 20.34}
  cardZone.setScale(_scale)
  updateZone()
  Global.call('objectLoadFinished',{self.guid})
end
function onCollisionEnter(collision)
  if cardZone ~= nil then
    updateZone()
  end
end
function updateZone()
  cardZone.setRotation(self.getRotation())
  local _position = self.getPosition()
  _position.y = 1.9
  cardZone.setPosition(_position)
end

function endPhase(o,c) board.call('controlClicked',{"endPhase",c}) end
function confirm(o,c) board.call('controlClicked',{"confirm",c}) end
function addOne(o,c) board.call('controlClicked',{"add",c}) end
function addTen(o,c) board.call('controlClicked',{"addTen",c}) end
function addAll(o,c) board.call('controlClicked',{"addAll",c}) end
function subOne(o,c) board.call('controlClicked',{"sub",c}) end
function subTen(o,c) board.call('controlClicked',{"subTen",c}) end
function subAll(o,c) board.call('controlClicked',{"subAll",c}) end
function card1(o,c) board.call('controlClicked',{"card",c,1}) end
function card2(o,c) board.call('controlClicked',{"card",c,2}) end
function card3(o,c) board.call('controlClicked',{"card",c,3}) end
 -- phase out
function updateTurn(value)
  data[1].label = value[1]
  data[1].color = pColors[value[1]]
  self.editButton(data[1])
end
 
function updatePhase(value)
  data[2].label = value[1]
  self.editButton(data[2])
end
function updateEndPhase(value)
  data[3].label = value[1]
  self.editButton(data[3])
end
function updateTerritoryOne(value)
  data[4].color = pColors[value[2]]
  data[4].font_color = tColors[value[2]]
  data[4].label = value[1]
  self.editButton(data[4])
end
function updateTerritoryTwo(value)
  data[5].color = pColors[value[2]]
  data[5].font_color = tColors[value[2]]
  data[5].label = value[1]
  self.editButton(data[5])
end
function updateConfirm(value)
  data[6].label = value[1]
  self.editButton(data[6])
end
function updateUnitsLeft(value)
  data[7].label = value[1]
  self.editButton(data[7])
end
function updateUnitsToPlace(value)
  data[12].label = value[1]
  self.editButton(data[12])
end
function updateUnitsLeftBehind(value)
  data[1].label = value[1]
  self.editButton(data[1])
end
function updateInfo(value)
  data[15].label = value[1]
  data[15].color = pColors[value[2]]
  data[15].font_color = tColors[value[2]]
  self.editButton(data[15])
end
function updateCardsButtons(value)
  local enabled = #value
  showCardZone({false})
  local disabled = 3 - enabled
  for i=1, enabled do
    local str = string.gsub(value[i], "%s+", "\n")
    data[15+i].label = str
    data[15+i].scale = {1,0.5,2}
    self.editButton(data[15+i])
  end
  for i=1,disabled do
    data[15+i+enabled].label = ""
    data[15+i+enabled].scale = {0,0,0}
    self.editButton(data[15+i+enabled])
  end
end
function showCardZone(value)
  if value[1] == true then
    data[22].scale = {1,0.5,2}
  else
    data[22].scale = {0,0,0}
  end
  self.editButton(data[22])
  
end
function doNothing(value)
end